rfd/0131-adminitrative-actions-mfa.md

```diff
@@ -0,0 +1,519 @@
+actions. We currently use this type of impersonation for our API and plugin
+guides, so each of these guides will need to be updated to use MachineID.
+
+### MFA enforcement
+
+MFA verification will be required to carry out various administrative actions.
+This will be enforced by the Auth API Server for the specific endpoints which
+are considered [administrative actions](#administrative-actions).
+
+Each Admin Action API request will follow this flow:
+
+1. Create an Auth API client using the user's existing valid certificates.
+1. Use the client to retrieve an MFA challenge for the user with
+   the existing `rpc CreateAuthenticateChallenge`.
+1. Prompt the user to solve the MFA challenge with their MFA device.
+1. Send the resulting `MFAAuthenticateResponse` to the Auth Server as part of the
+   API request.
+1. Validate and consume the `MFAAuthenticateResponse` in the Auth Server to
+   authorize the request (in addition to normal identity-based authorization).
+
+Steps 1-3 are already possible with Teleport currently and used for various MFA
+features. Steps 4 and 5 will require some changes to the Auth API client and
+server respectively.
+
+#### Server changes
+
+For admin actions, the Auth server will validate MFA for each request using the
+`MFAAuthenticateResponse` passed in the request.
+
+If the request fails MFA verification, an access denied error will be returned
+to the client.
+
+```go
+// ErrAPIMFARequired is returned by AccessChecker when an API request
+// requires an MFA check.
+var ErrAdminActionMFARequired = trace.AccessDenied("administrative action requires MFA")
+```
+
+#### Client changes
+
+There are a few different ways that the Auth client can pass MFA verification with
+the Auth server, each with their own pros and cons:
+
+1. Add the `MFAAuthenticateResponse` argument to each administrative API endpoint.
+
+The Auth server can consume/verify the challenge response in each request and
+treat the request as MFA verified.
+
+Pros:
+
+- It is straightforward and matches existing endpoints which require MFA
+  verification (rpc `GenerateUserSingleUseCerts`, login endpoints, etc.).
+- It is clear by the API reference which endpoints require MFA verification and
+  which do not, making it easier to reason from the client perspective, especially
+  for external users of the API.
+- The MFA verification only applies to one API request, preventing MFA-replay-like
+  attacks.
+
+Cons:
+
+- Requires custom changes to both the client and server implementations for
+  each affected endpoint, both to pass the MFA challenge response from the
+  client and for the Auth server to verify it for the request.
+- Old clients have no way to pass MFA until the `MFAAuthenticateResponse`
+  message is included in the proto specification. In order to maintain backwards
+  compatibility, each request that we change to an admin action will take a full
+  major version before we can actually require MFA for that action.
```

Comment:
I missed this issue before now. I update the backwards compatibility section to reflect this issue.

Essentially, each admin action will take a full major version before we can start enforcing MFA. For the MVP endpoints, we are planning to have done over Teleport 14+15. But for any endpoints that we add later in Teleport 14 or 15, we won't be able to require MFA until Teleport 16/17.

If we want to fix this, we could use a hybrid of Option 1 and 2. When a request is changed to require MFA, up to date clients and servers will know to pass the MFA response directly in the request. Older clients will make the request without MFA and receive `ErrAdminActionMFARequired`.  If the client is v14+, the client will retry by with MFA verification passed in the request metadata. As a result, as soon as an admin action is required on the server, v14+ clients will be able to meet the requirement.

WDYT? @r0mant @codingllama @rosstimothy

Reply:
Regardless of the approach, I think "new" clients need to know beforehand whether they need additional MFA for a call, so we can avoid the needless roundtrip. Ofc this varies per cluster settings, but we can make an educated guess. (Outdated clients will likely fail and then retry, but that seems alright.)

As for a 1+2) hybrid or just 2), I suppose having a single way to do things is easier to reason about. I'm failing right now to see how outdated clients, under option 1), could deal with `ErrAdminActionMFARequired` if they have an outdated proto.

Reply:
We'll be rolling out automatic upgrades for clients (tsh/tctl) in Q3 so this may be less of a concern but in general I agree that option 2 (having a generic method to passing MFA metadata rather than via API for each individual request) sounds like gives us a more graceful migration path.

Reply:
Ok, I think we should start with Option 2 then. It is the easier and more extensible solution. It should be possible to include all or most of the admin actions in Teleport 14, rather than only the current MVP list.

Once it is complete, we can consider going forward with option 1. This would give us a cleaner way to pass MFA in the long run, and for newly added requests. This is especially important for external users of the API as the metadata solution is not clear.

P.S. @codingllama this means we won't be splitting up existing gRPC requests into services that follow best practices, but we will still be making new gRPC services for affected HTTP endpoints.